File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.js
const fs = require("fs");

class LineReader {
  data = fs
    .readFileSync(process.platform === "linux" ? 0 : "input.txt", "utf-8")
    .toString()
    .trim()
    .split("\n");
  cursor = 0;

  read() {
    return this.data[this.cursor++];
  }

  readIntArray() {
    return this.read().split(" ").map(Number);
  }
}

const solve = () => {
  const lineReader = new LineReader();

  const [V, E] = lineReader.readIntArray();
  const edges = [];

  for (let i = 0; i < E; i++) {
    edges.push(lineReader.readIntArray());
  }

  const parents = Array.from({ length: V + 1 }, (_, i) => i);

  const find = (v) => {
    if (parents[v] === v) return v;
    return (parents[v] = find(parents[v]));
  };

  const union = (u, v) => {
    parents[find(u)] = find(v);
  };

  let count = 0;
  let sum = 0;

  edges.sort((a, b) => a[2] - b[2]);
    
  for (const [start, end, weight] of edges) {
    if (find(start) === find(end)) {
      continue;
    }
    union(start, end);
    count++;
    sum += weight;
    if (count === V - 1) {
      break;
    }
  }

  return sum;
};

console.log(solve());

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys
import heapq

input = sys.stdin.readline
vertex_count, edge_count = map(int, input().split())

edges = []
parent = [i for i in range(vertex_count + 1)]

for _ in range(edge_count):
    s, e, w = map(int, input().split())
    heapq.heappush(edges, (w, s, e))

def find(a):
    if parent[a] == a:
        return a
    parent[a] = find(parent[a])
    return parent[a]

def union(a, b):
    parent[find(a)] = find(b)

used_edge_count = 0
weight_sum = 0

while used_edge_count < vertex_count - 1:
    weight, start, end = heapq.heappop(edges)
    if find(start) != find(end):
        union(find(start), find(end))
        weight_sum += weight
        used_edge_count += 1

print(weight_sum)
